src/java/org/apache/cassandra/tcm/ClusterMetadata.java

```diff
@@ -0,0 +1,986 @@
+            return this;
+        }
+
+        public Transformer proposeRemoveNode(NodeId id)
+        {
+            tokenMap = tokenMap.unassignTokens(id);
+            return this;
+        }
+
+        public Transformer left(NodeId id)
+        {
+            tokenMap = tokenMap.unassignTokens(id);
+            directory = directory.withNodeState(id, NodeState.LEFT)
+                                 .withoutRackAndDC(id);
+            return this;
+        }
+
+        public Transformer with(DataPlacements placements)
+        {
+            this.placements = placements;
+            return this;
+        }
+
+        public Transformer with(LockedRanges lockedRanges)
+        {
+            this.lockedRanges = lockedRanges;
+            return this;
+        }
+
+        public Transformer with(InProgressSequences sequences)
+        {
+            this.inProgressSequences = sequences;
+            return this;
+        }
+
+        public Transformer with(ExtensionKey<?, ?> key, ExtensionValue<?> obj)
+        {
+            if (MetadataKeys.CORE_METADATA.contains(key))
+                throw new IllegalArgumentException("Core cluster metadata objects should be addressed directly, " +
+                                                   "not using the associated MetadataKey");
+
+            if (!key.valueType.isInstance(obj))
+                throw new IllegalArgumentException("Value of type " + obj.getClass() +
+                                                   " is incompatible with type for key " + key +
+                                                   " (" + key.valueType + ")");
+
+            extensions.put(key, obj);
+            modifiedKeys.add(key);
+            return this;
+        }
+
+        public Transformer withIfAbsent(ExtensionKey<?, ?> key, ExtensionValue<?> obj)
+        {
+            if (extensions.containsKey(key))
+                return this;
+            return with(key, obj);
+        }
+
+        public Transformer without(ExtensionKey<?, ?> key)
+        {
+            if (MetadataKeys.CORE_METADATA.contains(key))
+                throw new IllegalArgumentException("Core cluster metadata objects should be addressed directly, " +
+                                                   "not using the associated MetadataKey");
+            if (extensions.remove(key) != null)
+                modifiedKeys.add(key);
+            return this;
+        }
+
+        public Transformed build()
+        {
+            // Process extension first as a) these are actually mutable and b) they are added to the set of
+            // modified keys when added/updated/removed
+            for (MetadataKey key : modifiedKeys)
+            {
+                ExtensionValue<?> mutable = extensions.get(key);
+                if (null != mutable)
+                    mutable.withLastModified(epoch);
+            }
+
+            if (schema != base.schema)
```

Comment:
why do we compare all those things by references instead of values (like dumpDiff does)?